<a href="https://colab.research.google.com/github/micaeladepaolocabarcos/practicaproyectog/blob/main/U6_Limpieza_y_An%C3%A1lisis_Exploratorio_de_un_Dataset_Real.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Práctica: Limpieza y Análisis Exploratorio de un Dataset Real**
### Producción de pozos de gas y petróleo no convencional (Vaca Muerta, Argentina).
**Fuente:** Secretaría de Energía de la Nación — datos.gob.ar

https://colab.research.google.com/drive/1cHMxJVZXMJkEaACru-3h4jfjf-vtHfzX

####**1. Inicialización y carga**

In [ ]:
# importo las librerías que voy a usar
import pandas as pd
import numpy as np

# muestro todas las columnas cuando imprimo un DataFrame
pd.set_option('display.max_columns', 50)
# muestro los números con dos decimales para que sea más fácil leerlos
pd.set_option('display.float_format', '{:,.2f}'.format)

In [ ]:
# monto mi Google Drive para leer el CSV desde ahí
from google.colab import drive
drive.mount('/content/drive')

# defino la ruta del archivo
RUTA_CSV = '/content/drive/MyDrive/Coderhouse/produccion-de-pozos-de-gas-y-petrleo-no-convencional.csv'

In [ ]:
# cargo el dataset; uso low_memory=False para que pandas lea el archivo completo antes de inferir los tipos
df_original = pd.read_csv(RUTA_CSV, low_memory=False)

# trabajo sobre una copia para conservar el original intacto
df = df_original.copy()

print('Filas:', df.shape[0])
print('Columnas:', df.shape[1])

Filas: 421022
Columnas: 40


In [ ]:
# miro las primeras filas para conocer la estructura
df.head()

,idempresa,anio,mes,idpozo,prod_pet,prod_gas,prod_agua,iny_agua,iny_gas,iny_co2,iny_otro,tef,vida_util,tipoextraccion,tipoestado,tipopozo,observaciones,fechaingreso,rectificado,habilitado,idusuario,empresa,sigla,formprod,profundidad,formacion,idareapermisoconcesion,areapermisoconcesion,idareayacimiento,areayacimiento,cuenca,provincia,coordenadax,coordenaday,tipo_de_recurso,proyecto,clasificacion,subclasificacion,sub_tipo_recurso,fecha_data
0,YSUR,2016,1,137601,52.10,490.49,6.38,0.00,0.00,0.00,0.00,31.00,NaN,Surgencia Natural,Extracción Efectiva,Gasífero,NaN,2016-02-17 10:50:46.929347,f,t,5,YSUR ENERGÍA ARGENTINA S.R.L.,APA.RN.EFO-125(d),LAJA,"3,785.00",lajas,FEO,ESTACION FERNANDEZ ORO,Z155,ESTACION FERNANDEZ ORO,NEUQUINA,Rio Negro,-67.84,-39.02,NO CONVENCIONAL,GAS PLUS,EXPLOTACION,DESARROLLO,TIGHT,2016-01-31
1,YSUR,2018,1,158251,258.12,395.19,35.34,0.00,0.00,0.00,0.00,31.00,NaN,Surgencia Natural,Extracción Efectiva,Gasífero,NaN,2018-02-10 08:37:14.717426,f,t,444,YSUR ENERGÍA ARGENTINA S.R.L.,YEA.RN.EFO-121(d),LAJA,"4,505.00",lajas,FEO,ESTACION FERNANDEZ ORO,Z155,ESTACION FERNANDEZ ORO,NEUQUINA,Rio Negro,-67.86,-39.01,NO CONVENCIONAL,Sin Proyecto,EXPLOTACION,DESARROLLO,TIGHT,2018-01-31
2,YSUR,2017,1,157357,879.59,"1,753.20",176.69,0.00,0.00,0.00,0.00,31.00,NaN,Surgencia Natural,Extracción Efectiva,Gasífero,NaN,2017-02-16 13:45:37.233373,f,t,444,YSUR ENERGÍA ARGENTINA S.R.L.,YEA.RN.EFO-313(d),LAJA,"3,713.00",lajas,FEO,ESTACION FERNANDEZ ORO,Z155,ESTACION FERNANDEZ ORO,NEUQUINA,Rio Negro,-67.82,-39.01,NO CONVENCIONAL,Sin Proyecto,EXPLOTACION,DESARROLLO,TIGHT,2017-01-31
3,YSUR,2018,1,157882,571.08,991.33,9.09,0.00,0.00,0.00,0.00,31.00,NaN,Surgencia Natural,Extracción Efectiva,Gasífero,NaN,2018-02-10 08:37:14.717426,f,t,444,YSUR ENERGÍA ARGENTINA S.R.L.,YEA.RN.EFO-301(d),LAJA,"3,800.00",lajas,FEO,ESTACION FERNANDEZ ORO,Z155,ESTACION FERNANDEZ ORO,NEUQUINA,Rio Negro,-67.82,-39.01,NO CONVENCIONAL,Sin Proyecto,EXPLOTACION,DESARROLLO,TIGHT,2018-01-31
4,YSUR,2016,1,134008,8.88,325.92,20.14,0.00,0.00,0.00,0.00,31.00,NaN,Surgencia Natural,Extracción Efectiva,Gasífero,NaN,2016-02-17 10:50:46.929347,f,t,5,YSUR ENERGÍA ARGENTINA S.R.L.,AEA.RN.EFO-92,LAJA,"3,832.00",lajas,FEO,ESTACION FERNANDEZ ORO,Z155,ESTACION FERNANDEZ ORO,NEUQUINA,Rio Negro,-67.89,-39.01,NO CONVENCIONAL,GAS PLUS,EXPLOTACION,DESARROLLO,TIGHT,2016-01-31


####**2. Diagnóstico del estado de los datos**
*2.1 Tipos de datos y memoria*

In [ ]:
# reviso el tipo de cada columna, la cantidad de valores no nulos y el uso de memoria
df.info(memory_usage='deep')

<class 'pandas.DataFrame'>
RangeIndex: 421022 entries, 0 to 421021
Data columns (total 40 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   idempresa               421022 non-null  str    
 1   anio                    421022 non-null  int64  
 2   mes                     421022 non-null  int64  
 3   idpozo                  421022 non-null  int64  
 4   prod_pet                421022 non-null  float64
 5   prod_gas                421022 non-null  float64
 6   prod_agua               421022 non-null  float64
 7   iny_agua                421022 non-null  float64
 8   iny_gas                 421022 non-null  float64
 9   iny_co2                 421022 non-null  float64
 10  iny_otro                421022 non-null  float64
 11  tef                     421022 non-null  float64
 12  vida_util               8952 non-null    float64
 13  tipoextraccion          420408 non-null  str    
 14  tipoestado              420408 

Las fechas y los valores lógicos (`'t'`/`'f'`) están guardados como texto, y `vida_util` y `observaciones` tienen muy pocos datos. (En pandas 2.x, la versión de Colab, las columnas de texto figuran como `object`.)

*2.2 Estadística descriptiva de las variables numéricas principales*

In [ ]:
# selecciono las variables de producción, profundidad y tiempo efectivo
variables_numericas = ['prod_pet', 'prod_gas', 'prod_agua', 'profundidad', 'tef']

# calculo las estadísticas descriptivas
df[variables_numericas].describe()

,prod_pet,prod_gas,prod_agua,profundidad,tef
count,"421,022.00","421,022.00","421,022.00","421,022.00","421,022.00"
mean,333.70,629.71,188.77,"3,765.71",21.97
std,937.98,"1,628.89",625.46,"6,536.84",12.18
min,-0.00,-12.27,0.00,0.00,0.00
25%,0.00,8.23,0.00,"2,858.00",14.63
50%,7.79,103.64,12.65,"3,272.00",29.19
75%,122.31,449.96,86.48,"4,700.00",30.63
max,"26,593.26","40,086.80","34,792.95","378,939.00",79.34


Las medias son mucho mayores que las medianas (pocos pozos producen muchísimo) y aparecen valores imposibles: producciones negativas y profundidades de 0 m y de 378.939 m.

*2.3 Valores nulos*

Primero uso el método pedido en la consigna (`.isnull().sum()`) y después lo paso a porcentaje, para dimensionar el impacto de cada columna.

In [ ]:
# cuento los nulos por columna y me quedo solo con las que tienen alguno
nulos = df.isnull().sum()
nulos[nulos > 0]

vida_util           412070
tipoextraccion         614
tipoestado             614
tipopozo               614
observaciones       397033
clasificacion          922
subclasificacion       922
sub_tipo_recurso       452
dtype: int64

In [ ]:
def resumen_nulos(dataframe):
    # creo un diccionario vacío donde guardo el porcentaje de nulos de cada columna
    porcentajes = {}
    # guardo la cantidad total de filas
    total_filas = len(dataframe)

    # recorro cada columna del DataFrame
    for columna in dataframe.columns:
        # cuento los nulos de la columna
        cantidad = dataframe[columna].isnull().sum()
        # solo registro las columnas que tienen al menos un nulo
        if cantidad > 0:
            # calculo el porcentaje y lo guardo en el diccionario
            porcentajes[columna] = round(cantidad / total_filas * 100, 2)

    # convierto el diccionario en una Serie y la ordeno de mayor a menor
    return pd.Series(porcentajes, name='% nulos').sort_values(ascending=False)

resumen_nulos(df)

vida_util          97.87
observaciones      94.30
clasificacion       0.22
subclasificacion    0.22
tipopozo            0.15
tipoestado          0.15
tipoextraccion      0.15
sub_tipo_recurso    0.11
Name: % nulos, dtype: float64

`vida_util` y `observaciones` superan el 94% de nulos, así que las elimino; el resto tiene menos del 0,25% y se puede imputar o eliminar sin afectar el análisis.

*2.4 Duplicados*

Reviso dos tipos de duplicado: filas idénticas en todas sus columnas y filas repetidas según la clave lógica del dataset (un pozo reporta una sola vez por mes, así que `idpozo` + `anio` + `mes` debería ser única).

In [ ]:
# cuento las filas completamente duplicadas
print('Filas duplicadas (todas las columnas):', df.duplicated().sum())

# cuento las filas duplicadas según la clave pozo-año-mes
clave = ['idpozo', 'anio', 'mes']
print('Filas duplicadas (clave pozo-año-mes):', df.duplicated(subset=clave).sum())

Filas duplicadas (todas las columnas): 0
Filas duplicadas (clave pozo-año-mes): 0


No hay duplicados, pero dejo el paso `drop_duplicates()` en la limpieza por si la fuente (que se actualiza todos los meses) los incorpora.

*2.5 Valores inválidos o inconsistentes*

Además de nulos y duplicados, busco valores que existen pero no tienen sentido físico.

In [ ]:
# cuento producciones negativas (no puede haber producción menor a cero)
for columna in ['prod_pet', 'prod_gas', 'prod_agua']:
    print(f'{columna} negativos:', (df[columna] < 0).sum())

# cuento profundidades iguales a 0 (dato faltante cargado como cero) y mayores a 8000 m
print('profundidad = 0:', (df['profundidad'] == 0).sum())
print('profundidad > 8000:', (df['profundidad'] > 8000).sum())

prod_pet negativos: 1
prod_gas negativos: 2
prod_agua negativos: 0
profundidad = 0: 5738
profundidad > 8000: 152


In [ ]:
# miro qué valores aparecen en las profundidades mayores a 8000 m
df.loc[df['profundidad'] > 8000, 'profundidad'].value_counts().head()

profundidad
378,939.00    122
8,264.00       11
8,239.00       10
8,376.00        9
Name: count, dtype: int64

In [ ]:
# reviso las columnas de texto que en realidad representan valores lógicos (verdadero/falso)
for columna in ['rectificado', 'habilitado']:
    print(df[columna].value_counts(), '\n')

rectificado
f    420328
t       694
Name: count, dtype: int64 

habilitado
t    421022
Name: count, dtype: int64 



Detecto 3 producciones negativas, 5.738 profundidades en 0 y 152 mayores a 8.000 m (la mayoría con 378.939 m, un error de carga). Además, `habilitado` tiene un único valor, así que no aporta información.

**3. Limpieza rigurosa**

Aplico las transformaciones en un orden lógico: primero elimino lo que no aporta información, después corrijo los tipos (para que los cálculos posteriores sean confiables), luego trato nulos y valores inválidos y, por último, verifico duplicados y reseteo el índice.

*3.1 Eliminación de columnas sin información útil*

In [ ]:
# guardo la cantidad de filas y columnas antes de limpiar para comparar al final
filas_inicio = df.shape[0]
columnas_inicio = df.shape[1]

# elimino vida_util y observaciones (más del 94% nulos) y habilitado (tiene un único valor para todas las filas)
columnas_a_eliminar = ['vida_util', 'observaciones', 'habilitado']
df = df.drop(columns=columnas_a_eliminar)

print('Columnas eliminadas:', columnas_a_eliminar)
print('Columnas restantes:', df.shape[1])

Columnas eliminadas: ['vida_util', 'observaciones', 'habilitado']
Columnas restantes: 37


*3.2 Conversión de tipos de datos*

In [ ]:
# convierto las fechas de texto a datetime
df['fecha_data'] = pd.to_datetime(df['fecha_data'], format='%Y-%m-%d')
# fechaingreso tiene fracciones de segundo de largo variable, por eso uso el formato ISO8601
df['fechaingreso'] = pd.to_datetime(df['fechaingreso'], format='ISO8601')

# convierto rectificado de 't'/'f' a booleano con un diccionario de equivalencias
df['rectificado'] = df['rectificado'].map({'t': True, 'f': False})

# convierto a categoría las columnas de texto con pocos valores distintos que se repiten mucho
columnas_categoricas = ['idempresa', 'empresa', 'tipoextraccion', 'tipoestado', 'tipopozo',
                        'formprod', 'formacion', 'cuenca', 'provincia', 'tipo_de_recurso',
                        'proyecto', 'clasificacion', 'subclasificacion', 'sub_tipo_recurso']
for columna in columnas_categoricas:
    df[columna] = df[columna].astype('category')

# verifico los tipos de las columnas que modifiqué
df[['fecha_data', 'fechaingreso', 'rectificado', 'tipopozo', 'formacion']].dtypes

fecha_data      datetime64[us]
fechaingreso    datetime64[us]
rectificado               bool
tipopozo              category
formacion             category
dtype: object

*3.3 Imputación por moda del pozo: `tipopozo`, `tipoextraccion`, `tipoestado`*
Estas tres columnas tienen nulos en las mismas filas. En lugar de imputar con la moda general del dataset (que asignaría "Gasífero" a cualquier pozo), uso la moda de cada pozo: si un pozo tiene otros meses informados, completo con el valor que más se repite en su propio historial.

In [ ]:
# verifico que los nulos de las tres columnas caen en las mismas filas
nulos_tipopozo = df['tipopozo'].isnull()
print('Coinciden con tipoextraccion:', (nulos_tipopozo == df['tipoextraccion'].isnull()).all())
print('Coinciden con tipoestado:', (nulos_tipopozo == df['tipoestado'].isnull()).all())

Coinciden con tipoextraccion: True
Coinciden con tipoestado: True


In [ ]:
def moda_por_pozo(dataframe, columna):
    # creo un diccionario donde cada pozo tiene su propio contador de valores
    contadores = {}

    # recorro al mismo tiempo el id del pozo y el valor de la columna, fila por fila
    for pozo, valor in zip(dataframe['idpozo'], dataframe[columna]):
        # salteo los valores nulos porque no aportan a la moda
        if pd.isnull(valor):
            continue
        # si es la primera vez que veo este pozo, le creo un contador vacío
        if pozo not in contadores:
            contadores[pozo] = {}
        # si es la primera vez que veo este valor en el pozo, lo inicializo en cero
        if valor not in contadores[pozo]:
            contadores[pozo][valor] = 0
        # actualizo el contador
        contadores[pozo][valor] += 1

    # creo el diccionario final con la moda de cada pozo
    modas = {}
    # recorro cada pozo y su contador
    for pozo, conteo in contadores.items():
        # me quedo con el valor que tiene el conteo más alto
        modas[pozo] = max(conteo, key=conteo.get)

    return modas

In [ ]:
# recorro las tres columnas y completo sus nulos con la moda del pozo
for columna in ['tipopozo', 'tipoextraccion', 'tipoestado']:
    nulos_antes = df[columna].isnull().sum()
    # calculo la moda de cada pozo para esta columna
    modas = moda_por_pozo(df, columna)
    # busco la moda correspondiente a cada fila según su pozo
    valores_imputados = df['idpozo'].map(modas)
    # completo solo los nulos; paso a texto para poder combinar con la columna categórica
    df[columna] = df[columna].astype('object').fillna(valores_imputados).astype('category')
    nulos_despues = df[columna].isnull().sum()
    print(f'{columna}: {nulos_antes} nulos antes -> {nulos_despues} nulos después')

tipopozo: 614 nulos antes -> 4 nulos después


tipoextraccion: 614 nulos antes -> 4 nulos después


tipoestado: 614 nulos antes -> 4 nulos después


Recupero 610 de las 614 filas; las 4 restantes son de pozos sin ningún mes informado y las elimino en el paso siguiente.

*3.4 Eliminación estratégica de filas con nulos no recuperables*

Las filas que quedaron sin `tipopozo` y las que no tienen `clasificacion`, `subclasificacion` o `sub_tipo_recurso` pertenecen a pozos que no tienen ningún mes informado en esas columnas, por lo que no hay de dónde imputar. Como representan una proporción mínima del total, las elimino.

In [ ]:
# elimino las filas que siguen teniendo nulos en columnas categóricas clave
columnas_clave = ['tipopozo', 'tipoextraccion', 'tipoestado', 'clasificacion', 'subclasificacion', 'sub_tipo_recurso']
filas_antes = len(df)
df = df.dropna(subset=columnas_clave)
filas_eliminadas = filas_antes - len(df)

print('Filas eliminadas:', filas_eliminadas)
print(f'Porcentaje del total: {filas_eliminadas / filas_antes * 100:.2f}%')

Filas eliminadas: 1378
Porcentaje del total: 0.33%


*3.5 Eliminación de producciones negativas*

In [ ]:
# me quedo solo con las filas donde las tres producciones son mayores o iguales a cero
filas_antes = len(df)
df = df[(df['prod_pet'] >= 0) & (df['prod_gas'] >= 0) & (df['prod_agua'] >= 0)]
print('Filas eliminadas por producción negativa:', filas_antes - len(df))

Filas eliminadas por producción negativa: 3


*3.6 Imputación de `profundidad` inválida con la mediana por formación*

Trato como dato faltante tanto las profundidades en 0 (el pozo no puede tener profundidad nula) como las mayores a 8.000 m: el valor que más se repite ahí es 378.939 m, físicamente imposible (los pozos de Vaca Muerta rondan los 2.500 a 4.000 m de profundidad vertical), lo que indica un error de carga.

Como la profundidad de un pozo depende sobre todo de la formación geológica a la que apunta, imputo con la mediana de su formación. Elijo la mediana en lugar de la media porque es resistente a valores extremos.

In [ ]:
# reemplazo los valores inválidos por NaN para tratarlos como faltantes
df['profundidad'] = df['profundidad'].astype('float64')
invalidas = (df['profundidad'] == 0) | (df['profundidad'] > 8000)
df.loc[invalidas, 'profundidad'] = np.nan
print('Profundidades marcadas como faltantes:', df['profundidad'].isnull().sum())

Profundidades marcadas como faltantes: 5890


In [ ]:
def mediana_por_formacion(dataframe):
    # creo un diccionario donde guardo la lista de profundidades válidas de cada formación
    profundidades = {}

    # recorro al mismo tiempo la formación y la profundidad de cada fila
    for formacion, profundidad in zip(dataframe['formacion'], dataframe['profundidad']):
        # salteo las profundidades faltantes
        if pd.isnull(profundidad):
            continue
        # si es la primera vez que veo la formación, le creo una lista vacía
        if formacion not in profundidades:
            profundidades[formacion] = []
        # agrego la profundidad a la lista de su formación
        profundidades[formacion].append(profundidad)

    # calculo la mediana de cada formación
    medianas = {}
    for formacion, lista in profundidades.items():
        medianas[formacion] = np.median(lista)

    return medianas

# calculo las medianas y completo los faltantes según la formación de cada fila
medianas = mediana_por_formacion(df)
df['profundidad'] = df['profundidad'].fillna(df['formacion'].astype('object').map(medianas))
print('Profundidades faltantes después de imputar:', df['profundidad'].isnull().sum())

Profundidades faltantes después de imputar: 0


*3.7 Duplicados*

In [ ]:
# elimino duplicados exactos y por clave (en este dataset no hay, pero dejo el paso para que el flujo sea reproducible si el archivo se actualiza)
df = df.drop_duplicates()
df = df.drop_duplicates(subset=['idpozo', 'anio', 'mes'])
print('Duplicados por clave pozo-año-mes:', df.duplicated(subset=['idpozo', 'anio', 'mes']).sum())

Duplicados por clave pozo-año-mes: 0


*3.8 Reseteo del índice y verificación final*

In [ ]:
# reseteo el índice porque al eliminar filas quedaron huecos en la numeración
# uso drop=True para no guardar el índice viejo como una columna nueva
df = df.reset_index(drop=True)
df.index

RangeIndex(start=0, stop=419641, step=1)

In [ ]:
# verifico que las columnas numéricas sean realmente numéricas antes de hacer cálculos
for columna in ['prod_pet', 'prod_gas', 'prod_agua', 'profundidad', 'tef']:
    print(f'{columna}: {df[columna].dtype} -> numérica: {pd.api.types.is_numeric_dtype(df[columna])}')

prod_pet: float64 -> numérica: True
prod_gas: float64 -> numérica: True
prod_agua: float64 -> numérica: True
profundidad: float64 -> numérica: True
tef: float64 -> numérica: True


In [ ]:
# comparo el tamaño del dataset antes y después de la limpieza
print(f'Filas: {filas_inicio:,} -> {df.shape[0]:,} (se eliminaron {filas_inicio - df.shape[0]:,}, un {(filas_inicio - df.shape[0]) / filas_inicio * 100:.2f}%)')
print(f'Columnas: {columnas_inicio} -> {df.shape[1]}')

# confirmo que no quedan nulos
print('Nulos totales restantes:', df.isnull().sum().sum())

Filas: 421,022 -> 419,641 (se eliminaron 1,381, un 0.33%)
Columnas: 40 -> 37
Nulos totales restantes: 0


**4. Agregaciones de negocio**

**Unidades:** `prod_pet` y `prod_agua` están en m³; `prod_gas` en miles de m³.

### Pregunta 1: ¿Cómo evolucionó la producción anual total de petróleo y gas no convencional?

In [ ]:
# agrupo por año y sumo la producción de petróleo y gas
produccion_anual = df.groupby('anio')[['prod_pet', 'prod_gas']].sum()

# agrego la cantidad de pozos distintos que reportaron en cada año
produccion_anual['pozos_activos'] = df.groupby('anio')['idpozo'].nunique()
produccion_anual

,prod_pet,prod_gas,pozos_activos
anio,,,
2006,"15,067.28","425,467.59",199
2007,"20,821.97","414,711.99",207
2008,"39,629.87","699,831.65",250
2009,"65,410.50","1,009,414.97",146
2010,"66,063.40","1,230,472.88",195
2011,"144,142.20","1,833,805.90",244
2012,"303,274.99","2,391,052.62",340
2013,"555,595.45","2,916,657.69",530
2014,"1,121,582.49","4,942,706.07",876


La producción de petróleo pasa de unos 15 mil m³ en 2006 a casi 29,4 millones de m³ en 2025, con más de 20 veces más pozos activos. 2026 es un año incompleto (datos hasta julio).

### Pregunta 2: ¿Qué formaciones rinden más por pozo?
Calculo la producción mensual promedio por pozo en cada formación, separando pozos petrolíferos (petróleo) y gasíferos (gas), y considerando solo los meses en extracción efectiva para no promediar meses en que el pozo estuvo parado. Además, dejo afuera las formaciones con menos de 10 pozos, porque un promedio calculado sobre 1 o 2 pozos no es representativo.

In [ ]:
# me quedo con los meses en que el pozo estuvo produciendo
en_produccion = df[df['tipoestado'] == 'Extracción Efectiva']

# pozos petrolíferos: promedio y cantidad de pozos por formación
petroliferos = en_produccion[en_produccion['tipopozo'] == 'Petrolífero']
rendimiento_pet = petroliferos.groupby('formacion', observed=True).agg(
    prod_pet_promedio=('prod_pet', 'mean'),
    cantidad_pozos=('idpozo', 'nunique'))
# descarto las formaciones con menos de 10 pozos porque su promedio no es representativo
rendimiento_pet = rendimiento_pet[rendimiento_pet['cantidad_pozos'] >= 10]
rendimiento_pet.sort_values('prod_pet_promedio', ascending=False)

,prod_pet_promedio,cantidad_pozos
formacion,,
vaca muerta,913.50,2513
precuyo,120.62,10
lajas,117.84,29
agrio,110.68,14
quintuco,36.35,37
los molles,4.61,12


In [ ]:
# pozos gasíferos: promedio y cantidad de pozos por formación
gasiferos = en_produccion[en_produccion['tipopozo'] == 'Gasífero']
rendimiento_gas = gasiferos.groupby('formacion', observed=True).agg(
    prod_gas_promedio=('prod_gas', 'mean'),
    cantidad_pozos=('idpozo', 'nunique'))
# aplico el mismo criterio de al menos 10 pozos por formación
rendimiento_gas = rendimiento_gas[rendimiento_gas['cantidad_pozos'] >= 10]
rendimiento_gas.sort_values('prod_gas_promedio', ascending=False).head(6)

,prod_gas_promedio,cantidad_pozos
formacion,,
vaca muerta,"3,623.59",844
agrio,"2,213.06",55
punta rosada,"1,512.93",110
mulichinco,923.58,241
magallanes,691.59,185
lajas,532.58,581


Vaca Muerta es, por lejos, la formación más productiva por pozo: unos 914 m³ de petróleo por mes, más de 7 veces la siguiente, y también lidera en gas.

### Pregunta 3: ¿Qué empresas concentran la producción de petróleo?

In [ ]:
# sumo el petróleo producido por cada empresa en todo el período
petroleo_por_empresa = df.groupby('empresa', observed=True)['prod_pet'].sum().sort_values(ascending=False)

# calculo qué porcentaje del total representa cada empresa
participacion = (petroleo_por_empresa / petroleo_por_empresa.sum() * 100).round(2)

top_empresas = pd.DataFrame({'prod_pet_total': petroleo_por_empresa, 'participacion_%': participacion})
top_empresas.head(10)

,prod_pet_total,participacion_%
empresa,,
YPF S.A.,"81,798,102.29",58.24
VISTA ENERGY ARGENTINA SAU,"13,996,445.28",9.96
SHELL ARGENTINA S.A.,"9,797,705.78",6.98
PAN AMERICAN ENERGY SL,"7,505,730.73",5.34
PLUSPETROL S.A.,"4,922,509.86",3.50
TECPETROL S.A.,"4,188,210.55",2.98
VISTA OIL & GAS ARGENTINA SAU,"2,353,554.18",1.68
CHEVRON ARGENTINA S.R.L.,"2,213,027.33",1.58
EXXONMOBIL EXPLORATION ARGENTINA S.R.L.,"2,060,061.54",1.47


La producción está muy concentrada: YPF aporta el 58,2% del petróleo no convencional histórico y las 5 primeras empresas superan el 84%.

**5. Conclusiones**

Pasé de 421.022 filas y 40 columnas a 419.641 filas y 37 columnas, sin nulos, eliminando solo el 0,33% de las filas: la mayoría de los problemas se resolvieron imputando con criterio de dominio (moda del pozo y mediana de la formación). En lo analítico, la producción no convencional creció de forma sostenida, Vaca Muerta es la formación más productiva y la producción está muy concentrada en YPF.